File: src/main/java/ge/vakho/gxt/GXT.java
package ge.vakho.gxt;

import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.io.RandomAccessFile;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;
import java.util.Objects;
import java.util.TreeMap;

public class GXT {

    protected Block1 block1;
    protected Block2 block2;

    public static GXT from(Path path) throws IOException {
        return from(path.toFile());
    }

    public static GXT from(File file) throws IOException {
        try (RandomAccessFile raf = new RandomAccessFile(file, "r")) {
            return from(raf);
        }
    }

    public static GXT from(RandomAccessFile raf) throws IOException {
        GXT gxt = new GXT();
        gxt.setBlock1(Block1.from(raf));                // Block 1 (TKEY)
        gxt.setBlock2(Block2.from(raf, gxt.block1));    // Block 2 (TDAT)
        return gxt;
    }

    public static GXT from(Map<String, String> map) {
        GXT gxt = new GXT();
        gxt.block1 = Block1.from(map);          // Block 1 (TKEY)
        gxt.block2 = Block2.from(map.values()); // Block 2 (TDAT)
        return gxt;
    }

    public Map<String, String> toMap() {
        Map<String, String> result = new TreeMap<>();
        for (int i = 0; i < block1.tkeyEntries.size(); i++) {
            final TKEY tkey = block1.tkeyEntries.get(i);
            final TDAT tdat = block2.tdatEntries.get(i);
            result.put(tkey.tdatEntryName, tdat.text);
        }
        return result;
    }

    public void writeTo(File file) throws IOException {
        writeTo(file.toPath());
    }

    public void writeTo(Path path) throws IOException {
        try (OutputStream outputStream = Files.newOutputStream(path)) {
            writeTo(outputStream);
        }
    }

    public void writeTo(OutputStream outputStream) throws IOException {
        block1.writeTo(outputStream);
        block2.writeTo(outputStream);
    }

    public Block1 getBlock1() {
        return block1;
    }

    public void setBlock1(Block1 block1) {
        this.block1 = block1;
    }

    public Block2 getBlock2() {
        return block2;
    }

    public void setBlock2(Block2 block2) {
        this.block2 = block2;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        GXT gxt = (GXT) o;
        return Objects.equals(block1, gxt.block1) && Objects.equals(block2, gxt.block2);
    }

    @Override
    public int hashCode() {
        return Objects.hash(block1, block2);
    }

    @Override
    public String toString() {
        return "GXT{" +
                "block1=" + block1 +
                ", block2=" + block2 +
                '}';
    }

}

File: src/main/kotlin/ge/vakho/gxt/GXT.kt
package ge.vakho.gxt

import java.io.File
import java.io.IOException
import java.io.OutputStream
import java.io.RandomAccessFile
import java.nio.file.Files
import java.nio.file.Path
import java.util.*

/**
 * GXT consists of two blocks: [Block1] and [Block2] blocks. This class is responsible for parsing the GXT
 * information from a *.gxt file, using one of the overridden [from][.from] methods.
 *
 * This class also provides [toMap][.toMap] method to make working with GXT data easy. When you are done working
 * with the data, you may use [from][.from] method which accepts a [map][Map] object as an argument and
 * returns GXT object.
 *
 * To write GXT data to some file, or stream, use overridden [writeTo][.writeTo] methods.
 *
 * GTA III's GXT files are identical to those of GTA 2, except that they do not have headers – they start with the
 * TKEY block instead. There is also no special treatment of characters starting with 0x21 in TDAT entries.
 *
 * The Xbox version of Vice City uses this format as well.
 *
 * @see Block1
 * @see Block2
 */
data class GXT(var block1: Block1, var block2: Block2) {

    fun toMap(): Map<String, String> {
        val result = TreeMap<String, String>()
        var i = 0
        block1.entries.forEach { result[it.name] = block2.entries[i++].text }
        return result
    }

    @Throws(IOException::class)
    fun writeTo(file: File) {
        writeTo(file.toPath())
    }

    @Throws(IOException::class)
    fun writeTo(path: Path) {
        Files.newOutputStream(path).use { outputStream -> writeTo(outputStream) }
    }

    @Throws(IOException::class)
    fun writeTo(outputStream: OutputStream) {
        block1.entries.sort() // Sort TKEY entries before writing to file
        block1.writeTo(outputStream)
        block2.writeTo(outputStream)
    }

    companion object {
        @JvmStatic
        @Throws(IOException::class)
        fun from(path: Path): GXT {
            return from(path.toFile())
        }

        @JvmStatic
        @Throws(IOException::class)
        fun from(file: File): GXT {
            RandomAccessFile(file, "r").use { raf -> return from(raf) }
        }

        @JvmStatic
        @Throws(IOException::class)
        fun from(raf: RandomAccessFile): GXT {
            val block1 = Block1.from(raf)
            val block2 = Block2.from(raf, block1)
            return GXT(block1, block2)
        }

        @JvmStatic
        fun from(map: Map<String, String>): GXT {
            val block1 = Block1.from(map)
            val block2 = Block2.from(map.values)
            return GXT(block1, block2)
        }
    }
}